### DISCUSSION TITLE
Stat increases and decreases

### DISCUSSION THREAD

**Discussion by Sanglorian** [Tuxemon/Tuxemon#2523] [Ideas]
@JaskRendix has pointed out (#2512 ) that there are a lot of gaps in our stat-altering conditions. It got me thinking if the best solution might be to revisit an old proposal for temporary stat changes in combat. Then we can have techniques that alter stats in any way we like, without having to implement new conditions for them: 

* Stats can be adjusted in battle by "steps up" and "steps down". You can be anywhere between -6 and +6. 
* At step 0, the stat is multiplied by "3/3", in other words there is no change.
* For each step up, increase the numerator by one. So +1 means multiply by 4/3, +2 means multiply by 5/3, +3 means multiply by 6/3, and so on. 
* For each step down, increase the denominator by one. So -1 means multiply by 3/4, -2 means multiply by 3/5, and so on. 
* You cannot move beyond +6 steps or -6 steps. If you're on -3 and you would increase by +2, you go to -1.
* When the combat is over or when you swap out for another monster, all stat changes are wiped away. 

What do people think, especially @JaskRendix ? I think it could be a flexible way to accommodate lots of different stat changes. And the conditions you propose could become techniques that alter the stats in those ways. 

(Another thing I thought of: the Boost X items like Boost Ranged could increase by 2 steps rather than by 20%.)

**Comment by JaskRendix** [comment]
Interesting, from what I understood:
```
def calculate_adjusted_stat(base_stat, steps):
    steps = max(-6, min(6, steps))

    -----> these things to avoid division by zero = crash
    if steps == 6:
        adjusted_stat = base_stat * 9
    elif steps == -6:
        adjusted_stat = 0
    else:
        adjusted_stat = base_stat * (3 + steps) / (3 - steps)

    return round(adjusted_stat)
```
translated in numbers, considering a speed value of 100 and a melee value of 69 (both examples)
| Steps | Adjusted Stat (base 100) | Adjusted Stat (base 69) |
|---|---|---|
| -6 | 0.0 | 0.0 |
| -5 | 25.0 | 17.0 |
| -4 | 40.0 | 28.0 |
| -3 | 50.0 | 35.0 |
| -2 | 60.0 | 41.0 |
| -1 | 75.0 | 52.0 |
| 0 | 100.0 | 69.0 |
| 1 | 125.0 | 86.0 |
| 2 | 150.0 | 104.0 |
| 3 | 175.0 | 121.0 |
| 4 | 200.0 | 138.0 |
| 5 | 225.0 | 155.0 |
| 6 | 270.0 | 207.0 |

So, just to make sure I understand how this works - if we have a technique like '**goad**' with effects like '**step_up melee**' and '**step_down ranged**' (for the **enraged** one).

For example, if '**step_up**' is melee and '**step_down**' is ranged, and the values are 100 for both, that would mean melee increases to 125 and ranged decreases to 75. But here's my question: what happens if the monster uses 'goad' again? Would melee go up to 150 and ranged go down to 60? Same question for the item, if I use boost XXX twice, same?

note: I've added a base value of 69 to the example, so we can see how the system handles uneven numbers and make sure everything is working as expected

**Reply by Sanglorian** [reply]
Sorry I forgot to account for the negative numbers in calculating the fraction. This is what I had in mind:

`
def calculate_adjusted_stat(base_stat, steps):
    steps = max(-6, min(6, steps)

    if steps > 0:
        adjusted_stat = base_stat * (3 + steps) / 3
    else:
        adjusted_stat = base_stat * (3) / (3 - steps)

    return round(adjusted_stat)
`
So with 100 and 69 stats, 

| Steps | 100 | 69  |
|-------|-----|-----|
| -6    | 33  | 23  |
| -5    | 38  | 26  |
| -4    | 43  | 30  |
| -3    | 50  | 35  |
| -2    | 60  | 41  |
| -1    | 75  | 52  |
| 0     | 100 | 69  |
| 1     | 133 | 92  |
| 2     | 167 | 115 |
| 3     | 200 | 138 |
| 4     | 233 | 161 |
| 5     | 267 | 184 |
| 6     | 300 | 207 |

> So, just to make sure I understand how this works - if we have a technique like 'goad' with effects like 'step_up melee' and 'step_down ranged' (for the enraged one).

I would do it as "step_up melee 1" or "step_down ranged 2", so a technique can change things by more than one step at a time. And then yes, an effect could work multiple times with additive effect - two uses of step_up melee 1 would mean +2 steps for melee. 

And then I think it's an interesting question what to do with conditions that alter stats ... It might be cleaned to get rid of them altogether if that's the only thing they so. Interested to hear what others think though!

**Reply by JaskRendix** [reply]
ok, "step_up" and "step_down" are complete. I'm going to set up some additional testing. So far, the tests with Pairagrin have gone as expected. I ran into some trouble with negative values, but it was just a simple issue with brackets. Working on the items (boosters) too.

some `step_up`  (by default +1 each time) examples:
`step_up user,speed`
`step_up target,speed`
`step_up both,speed`
`step_up user,speed:melee:dodge`
same with `step_down` (by default -1 each time)

**Reply by Sanglorian** [reply]
Thank you! This opens up a lot of options for techniques!

**Reply by Sanglorian** [reply]
Been thinking about conditions. I do have a soft spot for Sniping and Enraged. Those are quite specific and I like the way they involve trade offs. Sometimes it's a benefit to get the condition and sometimes it's a cost. Charged up and Exhausted as well are very specific and limited in time.

But I think focused, slow, hard shell, softened and blinded could potentially all be removed and replaced by stat changes under this system.

What do people think?

**Reply by JaskRendix** [reply]
We can keep everything as is, including the icons, but what we need to figure out is how to effectively communicate the increase or decrease of statistics to the player without relying on the condition icons. One possible solution could be to use a standard message that clearly indicates the change, so the player knows what's happening without needing a specific condition icon.

**Reply by Sanglorian** [reply]
Good point, thank you! 

We could do: 
* TUXEMON's STAT fell by 1 step. 
* TUXEMON's STAT raised by 2 steps.
* etc

If we wanted to do icons, a simple one would be sprites of +1, +2, etc and -1, -2, etc. 

We have associated colours thanks to the stat boost items: blue for Armor, yellow for Dodge, red for Melee, green for Ranged, and orange/brown for Speed. (Although the items vary a bit between Boost X and Raise X).

These could show the final result (so showing a green icon of +1 if they received +2 steps Ranged and -1 step Ranged).

**Reply by HippasusTwo** [reply]
I really like this idea of being able to lower and raise stats. It makes using a high power, long cooldown move something you can potentially counter-strategy by taking the opportunity to set up if the opponent doesn't have good moves that can take you out before then.
